tensorflow/compiler/mlir/xla/transforms/xla_hlo_fusion.cc

```diff
@@ -0,0 +1,553 @@
+          for (int output2 = output1+1; output2 < numOutput; ++output2)
+            if (op_fusibility.inferOutputsShapeEquality(output1, output2))
+              update(op->getResult(output1), op->getResult(output2));
+
+        // shape equality propagation between input and output.
+        for (int input = 0; input < numInput; ++input)
+          for (int output = 0; output < numOutput; ++output)
+            if (op_fusibility.inferInputOutputShapeEquality(input, output))
+              update(op->getOperand(input), op->getResult(output));
+      }
+    } while (!converged);
+  }
+
+  // a UnionFind set
+  EquivalenceClasses<Value> impl_;
+};
+
+// A fusion planner that can propose a fusion plan for a block of ops.
+// The fusion plan is consisted of a group of fusion patterns.
+//
+// Currently all proposed patterns followed xla kLoop/kInput like fusion
+// templates while are adapted to the fully dynamic shape world.
+//
+// kLoop fusion template satifies:
+//   - all ops in the fusion pattern are element-wise.
+//   - all the shapes of outputs of fusion pattern are same, and thus can
+//     fit into a same parallel loop.
+//
+// kInput fusion template satifies:
+//   - any op in the fusion pattern is either element-wise or a reduction.
+//   - if a op is a reduction, its output cannot be consumered by other
+//     ops in the same fusion pattern.
+//   - all the effective shapes of outputs of fusion pattern are same.
+//     - For element-wise op, its effective shape is its output shape.
+//     - For reduction op, its effective shape is its operand shape.
+class FusionPlanner {
+ public:
+  explicit FusionPlanner(const SmallVectorImpl<Operation *>& op_list)
+   : op_list_(op_list), shape_analysis_(op_list) {
+    BuildNodeMap();
+  }
+
+  // Return a fusion plan if success, otherwise none.
+  llvm::Optional<FusionPlan> Run() {
+    // Greedily search connected fusible pattern, and ops belonging to
+    // a same fusion pattern are grouped into a cluster.
+    RunEdgeContractionLoop();
+
+    // After doing edge contraction, each unique cluster having size
+    // more than one represents a potential fusion pattern.
+    // We collect all these clusters and construct a fusion plan.
+    //
+    // Note that the ops in a fusion pattern are in topological ordering.
+    FusionPlan plan;
+    DenseMap<int, int> pattern_ids;
+    for (Operation* op : op_list_) {
+      Cluster* cluster = GetClusterForNode(op);
+      int node_id = cluster->cycles_graph_node_id();
+      if (!IsFusible(op_list_[node_id]) ||
+          EffectiveSize(GetClusterForNode(op)->fused_pattern()) <= 1) {
+        continue;
+      }
+      if (!pattern_ids.count(node_id)) {
+        int pattern_id = pattern_ids.size();
+        pattern_ids[node_id] = pattern_id;
+        plan.emplace_back();
+      }
+      plan[pattern_ids[node_id]].push_back(op);
+    }
+    return plan;
+  }
+
+  // Return the op_list this planner operates on.
```

Comment:
Returns: 'These comments should open with descriptive verbs in the indicative mood ("Opens the file") rather than verbs in the imperative ("Open the file")"'  https://google.github.io/styleguide/cppguide.html#Function_Comments